### PR TITLE
Add Gitlab/Github feedback link icon in Techdocs Reader

### PR DESCRIPTION
**Pull request by adrianke77**
Signed-off-by: Chongyang Adrian, Ke <ftt.adrian.ke@grabtaxi.com>

## Hey, I just made a Pull Request!

Add feedback link icon in Techdocs Reader that directs to Gitlab or Github repo issue page with prefilled title and source link.

For link to appear, requires repo_url and edit_uri to be filled in mkdocs.yml, as per https://www.mkdocs.org/user-guide/configuration

Icon is on right, speech box with exclamation mark. Edit icon is added by MkDocs native functionality. Feedback icon only appears when Edit icon is available.

image

Gitlab issue page with prefilled info:
image

Github issue page with prefilled info:
image

#### :heavy_check_mark: Checklist

- [ :heavy_check_mark: ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ NA ] Added or updated documentation
- [ :heavy_check_mark: ] Tests for new functionality and regression tests for bug fixes
- [ :heavy_check_mark: ] Screenshots attached (for UI changes)
- [ :heavy_check_mark: ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      gitHost = 'gitlab';
+    } else if (sourceAnchor.href.includes('github')) {
+      gitHost = 'github';
+    } else {
+      return dom;
+    }
+
+    const title = (contentContainer.querySelector(
+      'h1',
+    ) as HTMLElement).textContent!.replace(/[^a-zA-Z ]/g, '');
```
Comment (OrkoHunter) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:44]: If possible, please add a comment here explaining what's happening.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    ) as HTMLAnchorElement;
+    // don't show if Gitlab or Github hosting not specified
+    if (!sourceAnchor || !sourceAnchor.href) {
+      return dom;
+    }
+    let gitHost = '';
+
+    if (sourceAnchor.href.includes('gitlab')) {
+      gitHost = 'gitlab';
+    } else if (sourceAnchor.href.includes('github')) {
```
Comment (OrkoHunter) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:36]: We should highlight that this will not work for ghe or any self-hosted GitHub/GitLab where the host does not contain the string `gitlab` or `github`. e.g. `ghe.mycompany.net` or `git.mycompany.com`.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        feedbackLink.href = `${sourceURL.origin}${repoPath}/issues/new?title=${issueTitle}&body=${issueDesc}`;
+        break;
+      default:
+        return dom;
+    }
+    feedbackLink.insertAdjacentHTML(
+      'afterbegin',
+      // 'Feedback' icon from https://material-ui.com/components/material-icons
+      '<svg <g style="width:22px"><path d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-7 12h-2v-2h2v2zm0-4h-2V6h2v4z"></path></g></svg>',
+    );
```
Comment (OrkoHunter) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:67]: Oooh, a bit sad that we can't do this by importing the icon and not copying the svg. Maybe there is a way?

```
import FeedbackIcon from '@material-ui/icons/Feedback';
```

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    feedbackLink.title = 'Leave feedback for this page';
+    feedbackLink.id = 'git-feedback-link';
+    sourceAnchor?.insertAdjacentElement('beforebegin', feedbackLink);
+    return dom;
+  };
+};
+
+// gitlab.myteksi.net/nexus/docs/-/tree/master/confluence-convert-test/conveyor/docs/Conveyor.md
+
+// https: `https://gitlab.com/gitlab-org/gitlab-ce/issues/new?issue[title]=Validate%20new%20concept&issue[description]=Research%20idea`;
```
Comment (OrkoHunter) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:78]:
```suggestion
```

**Review by OrkoHunter — COMMENTED** [review]
Nice - a decent start for https://github.com/backstage/backstage/issues/2032 I think.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 2 above)
Comment (Fox32) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:36]: You could use `giturlparse` to detect self hosted instances.

Comment (Fox32) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:36]: Or the @backstage/integrations package

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 Spotify AB
```
Comment (emmaindal) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:2]:
```suggestion
 * Copyright 2021 Spotify AB
```

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * limitations under the License.
+ */
+
+import type { Transformer } from './index';
+
+// requires repo
+export const addGitFeedbackLink = (): Transformer => {
+  return dom => {
+    const contentContainer = dom.querySelector(
+      '.md-content__inner',
```
Comment (emmaindal) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:23]: The risk with relying too much on MkDocs material classes is that this will break as soon as we upgrade to a new version that possibly include changes to their classes (or someone are using an older/newer version that does not have this specific class). It would be nice if we can have power over this feature to 100% without relying on classes from mkdocs material.

Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:23]: Could try using non-class identifiers, i'll see what can be done.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 3 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:67]: Hmm perhaps material-ui/icons exposes the svg somewhere. I'll take a look

File: .changeset/techdocs-eight-camels-hear.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-techdocs': minor
+---
+
+Add feedback link icon in Techdocs Reader that directs to Gitlab or Github repo issue page with prefilled title and source link.
```
[the remark this answers is not in the document]
Comment (adrianke77) [reply .changeset/techdocs-eight-camels-hear.md:5]: I haven't checked, i'll take a look at Bitbucket's system.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 1 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:44]: Updated

File: .changeset/techdocs-eight-camels-hear.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (adrianke77) [reply .changeset/techdocs-eight-camels-hear.md:5]: I don't seem to be able to find the same functionality for Bitbucket i.e. prefill Bitbucket issue form by URL. 

I suggest this be left to another contributor with access to a Bitbucket instance to implement, as I don't have access to one at this time.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 2 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:36]: I'm stumped. For URLs that don't include the name of the implementation as a string, not sure if we can discern without additional information. `giturlparse` doesn't specify the provider, and the docs may be built externally and hence use a Git hosting solution that isn't set up in Backstage's integrations config.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 1 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts]: Changed selector to .childNodes[0].textContent to avoid the regex, which was to strip the link character from title string

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 2 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts]: I've tried to add keys in app-config at techdocs.feedback.gitlab and .github for matching as Gitlab/Github Git hosting types, but having some problems making values added there appear in the config object. Perhaps there is some schema or config getter i need to modify that I'm unable to find so far?

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 3 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts]: Updated

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 2 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts]: Fixed above. There are now optional keys in app-config under techdocs.feedback.gitlab and .github for specifying hostnames that should be identified as Gitlab or Github hosting.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts (hunk 6 above)
Comment (adrianke77) [reply plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts]: Updated

**Comment by adrianke77** [comment]
Had to squash and fix commits to revert some yarn.lock change that was somehow introducing Typescript errors due to updating of some packages.

File: plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+ * limitations under the License.
+ */
+
+import type { Transformer } from './index';
+import FeedbackOutlinedIcon from '@material-ui/icons/FeedbackOutlined';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+// requires repo
+export const addGitFeedbackLink = (configApi: any): Transformer => {
```
Comment (adrianke77) [inline plugins/techdocs/src/reader/transformers/addGitFeedbackLink.ts:23]: Had difficulty finding a type to use for configApi here, any advice would be appreciated for replacing the 'any'.

File: .changeset/techdocs-eight-camels-hear.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+For link to appear, requires repo_url and edit_uri to be filled in mkdocs.yml, as per https://www.mkdocs.org/user-guide/configuration.
```
Comment (realandersn) [inline .changeset/techdocs-eight-camels-hear.md:6]: edit_uri does not seem to be required for GitHub, Bitbucket, and GitLab (at least for master branch for GitHub and GitLab, and default branch for Bitbucket) as per the note in https://www.mkdocs.org/user-guide/configuration/#edit_uri

**Comment by realandersn** [comment]
General comment: would be nice to not render the icon when issues are disabled on the repo, as it will now lead to a 404 link in that case. This could of course also be iterated upon at a later stage.

File: .changeset/techdocs-eight-camels-hear.md (hunk 9 above)
Comment (adrianke77) [reply .changeset/techdocs-eight-camels-hear.md:6]: edit_uri apparently needs repo_url to have the common hostnames in order to be set automatically (eg gitlab.com, github.com). Self-hosting with other hostnames need a edit_uri specified.

Comment (adrianke77) [reply .changeset/techdocs-eight-camels-hear.md]: Added note on above.

File: plugins/techdocs/config.d.ts [hunk 10]
```diff
@@ -32,5 +32,19 @@ export interface Config {
      * @deprecated
      */
     requestUrl?: string;
+
+    /**
+     * Alternate hostnames to identify Git hosting type for feedback links for TechDocs pages
+     */
+    feedback?: {
```
Comment (iamEAP) [inline plugins/techdocs/config.d.ts:39]: What do you think of something like...

```typescript
export interface Config {
  techdocs: {
    /**
     * Used to identify source code or issue system for feedback links for TechDocs pages
     */
    feedback?: {
      /**
       * The host name of the target system (e.g. ghe.example.com)
       * @visibility frontend
       */
      hostName?: string;
      /**
       * Name of the target system (e.g. github)
       * @visibility frontend
       */
      provider?: string;
    };
  };
}
```

Would prevent invalid configs (e.g. specifying both `techdocs.feedback.github` and `techdocs.feedback.gitlab`), as well as keep the config schema from growing with more supported providers.

**Review by iamEAP — COMMENTED** [review]
Thanks @adrianke77!  Works beautifully.  A suggestion on the config schema.

File: plugins/techdocs/config.d.ts (hunk 10 above)
Comment (adrianke77) [reply plugins/techdocs/config.d.ts:39]: Specifying multiple feedback hostnames still works, it just uses the first match available. This is in case of documents coming from different providers needing to be displayed in the same Backstage instance.

Comment (iamEAP) [reply plugins/techdocs/config.d.ts:39]: 🤔 &nbsp;Oh, interesting.  I guess I hadn't fully thought through the prospect that one's catalog might be made up of entities whose code is stored across multiple source code providers.

Given that's the case, maybe we should more closely mirror the way `integrations`' config is implemented (where each provider key is an array with details)?

Or actually, can we just pull these details from the `integrations` config itself?

Comment (adrianke77) [reply plugins/techdocs/config.d.ts:39]: Will change to use `integrations`

Comment (iamEAP) [reply plugins/techdocs/config.d.ts]: 🤦‍♂️  As I was reading through / closing out past comments, I realized you @adrianke77  had made a comment about `integrations`...  Sorry for missing that.  To look at that one more time:

> _@backstage/integrations also won't be of help if the built HTML docs are from a Git source not specified in the integrations configuration, eg if they are built on external CI/CD._

I suppose the case where the `integrations` config won't be populated will be somewhat uncommon?  Even if one sets up TechDocs in the recommended way (with publishing via CI/CD), it's likely you still have one or more integrations configured for other features in Backstage (e.g. for pulling component yaml definitions from the Catalog, or for generating new components via the Scaffolder).

File: .changeset/techdocs-eight-camels-hear.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+---
+'@backstage/plugin-techdocs': minor
+---
+
+Add feedback link icon in Techdocs Reader that directs to Gitlab or Github repo issue page with prefilled title and source link.
+For link to appear, requires repo_url and edit_uri to be filled in mkdocs.yml, as per https://www.mkdocs.org/user-guide/configuration. edit_uri will need to be specified for self-hosted Gitlab/Github with different hostnames.
+Gitlab or Github detection checks for 'gitlab' or 'github' in hostname of source in repo_url;
+additionally 'techdocs.feedback.gitlab' and '.github' in app config can be used to specify hostnames that should be recognised as the respective Git hosting for the feedback link.
```
Comment (iamEAP) [inline .changeset/techdocs-eight-camels-hear.md:8]: Should be updated to something like...

```suggestion
additionally, the 'host' key on 'gitlab' and 'github' configurations in the app 'integrations' config will be used to determine the feedback link.
```

**Review by iamEAP — COMMENTED** [review]
Thanks @adrianke77!  Quick docs updates and that test fix and I think we should be good to go.

It's also probably worth adding a note about this in the general TechDocs documentation too...  There isn't a great spot for it currently (there's no documentation on how to "consume" / "view" docs), so maybe just in the [FAQ section](https://github.com/backstage/backstage/blob/master/docs/features/techdocs/FAQ.md) for now?  Here's some sample copy:

```md
#### Is it possible for users to suggest changes or provide feedback on a TechDocs page?
This is supported for TechDocs sites whose source code is hosted in either GitHub or GitLab.  In order to add "edit this page" and "leave feedback" buttons on a TechDocs page, be sure that you have `repo_url` and `edit_uri` values in your `mkdocs.yml` files per [MkDocs instructions](https://www.mkdocs.org/user-guide/configuration), and that you have an `integrations` entry in your `app-config.yaml` pointed at your source code provider (only the `host` key is necessary).
```

File: plugins/techdocs/config.d.ts (hunk 10 above)
Comment (adrianke77) [reply plugins/techdocs/config.d.ts]: Yeah will ignore that edge case on the grounds that it seems quite unlikely, for now.

**Comment by adrianke77** [comment]
> Thanks @adrianke77! Quick docs updates and that test fix and I think we should be good to go.
> 
> It's also probably worth adding a note about this in the general TechDocs documentation too... There isn't a great spot for it currently (there's no documentation on how to "consume" / "view" docs), so maybe just in the [FAQ section](https://github.com/backstage/backstage/blob/master/docs/features/techdocs/FAQ.md) for now? Here's some sample copy:
> 
> ```lisp
> #### Is it possible for users to suggest changes or provide feedback on a TechDocs page?
> This is supported for TechDocs sites whose source code is hosted in either GitHub or GitLab.  In order to add "edit this page" and "leave feedback" buttons on a TechDocs page, be sure that you have `repo_url` and `edit_uri` values in your `mkdocs.yml` files per [MkDocs instructions](https://www.mkdocs.org/user-guide/configuration), and that you have an `integrations` entry in your `app-config.yaml` pointed at your source code provider (only the `host` key is necessary).
> ```

Updated

**Review by iamEAP — APPROVED** [review]

**Comment by iamEAP** [comment]
Thanks again @adrianke77!  Just had to satisfy vale with a couple of quick changes to docs.  Excited to have this in TechDocs!
